crates/copilot/src/copilot_chat.rs

```diff
@@ -68,6 +90,9 @@ impl Model {
             Self::Gpt4o => 128000,
             Self::Gpt4 => 8192,
             Self::Gpt3_5Turbo => 16385,
+            Self::O1Mini => 128000,
+            Self::O1Preview => 128000,
+            Self::Claude3_5Sonnet => 200_000,
```

Comment:
@overjt where these numbers are coming from?

Reply:
Those were the limits when I did the implementation (hours after GitHub's announcement), apparently the limits were changed in recent days.

The models and limits can be consulted in the copilot api
`https://api.individual.githubcopilot.com/models`

```json
{
  "data": [
    {
      "capabilities": {
        "family": "gpt-3.5-turbo",
        "limits": {
          "max_context_window_tokens": 16384,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-3.5-turbo",
      "model_picker_enabled": false,
      "name": "GPT 3.5 Turbo",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-3.5-turbo-0613"
    },
    {
      "capabilities": {
        "family": "gpt-3.5-turbo",
        "limits": {
          "max_context_window_tokens": 16384,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-3.5-turbo-0613",
      "model_picker_enabled": false,
      "name": "GPT 3.5 Turbo",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-3.5-turbo-0613"
    },
    {
      "capabilities": {
        "family": "gpt-4",
        "limits": {
          "max_context_window_tokens": 32768,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 32768
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-4",
      "model_picker_enabled": false,
      "name": "GPT 4",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4-0613"
    },
    {
      "capabilities": {
        "family": "gpt-4",
        "limits": {
          "max_context_window_tokens": 32768,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 32768
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-4-0613",
      "model_picker_enabled": false,
      "name": "GPT 4",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4-0613"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o",
      "model_picker_enabled": true,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-05-13"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-2024-05-13",
      "model_picker_enabled": false,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-05-13"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4-o-preview",
      "model_picker_enabled": false,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-05-13"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 16384,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-2024-08-06",
      "model_picker_enabled": false,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-08-06"
    },
    {
      "capabilities": {
        "family": "o1",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-preview",
      "model_picker_enabled": true,
      "name": "o1-preview (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-preview-2024-09-12"
    },
    {
      "capabilities": {
        "family": "o1",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-preview-2024-09-12",
      "model_picker_enabled": false,
      "name": "o1-preview (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-preview-2024-09-12"
    },
    {
      "capabilities": {
        "family": "o1-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-mini",
      "model_picker_enabled": true,
      "name": "o1-mini (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-mini-2024-09-12"
    },
    {
      "capabilities": {
        "family": "o1-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-mini-2024-09-12",
      "model_picker_enabled": false,
      "name": "o1-mini (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-mini-2024-09-12"
    },
    {
      "capabilities": {
        "family": "text-embedding-ada-002",
        "limits": {
          "max_inputs": 256
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "cl100k_base",
        "type": "embeddings"
      },
      "id": "text-embedding-ada-002",
      "model_picker_enabled": false,
      "name": "Embedding V2 Ada",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "text-embedding-ada-002"
    },
    {
      "capabilities": {
        "family": "text-embedding-3-small",
        "limits": {
          "max_inputs": 512
        },
        "object": "model_capabilities",
        "supports": {
          "dimensions": true
        },
        "tokenizer": "cl100k_base",
        "type": "embeddings"
      },
      "id": "text-embedding-3-small",
      "model_picker_enabled": false,
      "name": "Embedding V3 small",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "text-embedding-3-small"
    },
    {
      "capabilities": {
        "family": "text-embedding-3-small",
        "object": "model_capabilities",
        "supports": {
          "dimensions": true
        },
        "tokenizer": "cl100k_base",
        "type": "embeddings"
      },
      "id": "text-embedding-3-small-inference",
      "model_picker_enabled": false,
      "name": "Embedding V3 small (Inference)",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "text-embedding-3-small"
    },
    {
      "capabilities": {
        "family": "gpt-4o-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-mini",
      "model_picker_enabled": false,
      "name": "GPT 4o Mini",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-mini-2024-07-18"
    },
    {
      "capabilities": {
        "family": "gpt-4o-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-mini-2024-07-18",
      "model_picker_enabled": false,
      "name": "GPT 4o Mini",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-mini-2024-07-18"
    },
    {
      "capabilities": {
        "family": "gpt-4-turbo",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-4-0125-preview",
      "model_picker_enabled": false,
      "name": "GPT 4 Turbo",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4-0125-preview"
    },
    {
      "capabilities": {
        "family": "claude-3.5-sonnet",
        "limits": {
          "max_context_window_tokens": 200000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 195000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "claude-3.5-sonnet",
      "model_picker_enabled": true,
      "name": "Claude 3.5 Sonnet (Preview)",
      "object": "model",
      "policy": {
        "state": "enabled",
        "terms": "Enable access to the latest Claude 3.5 Sonnet model from Anthropic. [Learn more about how GitHub Copilot serves Claude 3.5 Sonnet](https://docs.github.com/copilot/using-github-copilot/using-claude-sonnet-in-github-copilot)."
      },
      "preview": true,
      "vendor": "Anthropic",
      "version": "claude-3.5-sonnet"
    }
  ],
  "object": "list"
}```